### DISCUSSION TITLE
Refactor Cline's configuration design (sooner than later)

### DISCUSSION THREAD

**Discussion by KJ7LNW** [cline/cline#835] [Feature Requests]
PR #784 has shown that Cline has outgrown it's current configuration design. 

In #784's diff, you can see that, just in terms of adding flags and initialization (_without_ including actual implementation) that there are 12+ new lines of code that are required **just to add a _single_ configuration flag**! 

Cline is getting popular and others are going to extend its configuration, too, so refactoring a common configuration framework should happen sooner than later. Certainly there are many ways to simplify configuration flags with a framework so that they can be added in one location and propagate to everywhere that might need it. 

I am new to TS and I do not know the language best practices are, so here is just one example of structuring a configuration description:

```perl
    # the configuration might be defined globally as something like this:
    config_desc =>
        { 
            customInstructions => {
                type => string,
                description => "your custom instructions",
                default => "",
                css => 'textarea',
            },
            alwaysAllowWrite => {
                type => boolean,
                default => false,
                css => 'checkbox',
            },
            ...
        }
```

A configuration object can then be passed around, which is derived from the description above; it might need to be a class with values, but to show this has a structure for proof of concept it might look like this:

```perl
    config =>
        {
            customInstructions => foo,
            alwaysAllowWrite => true,
            ...
        }
```

and pass the configuration around _instead_ of having to patch configuration options all over the place every time a single option is added. For example, generalize all `const { ... } = await this.getState()` calls to simply return the entire config.

These are just some of the places that need generalization:

- `class Cline` initializers. 
```diff
-this.cline = new Cline(this, apiConfiguration, customInstructions, alwaysAllowReadOnly, task, images)
+this.cline = new Cline(this, config, task, images)
```

- if this is not taken care of soon, instantiation is going to look crazy when #784 is merged: 

```ts
this.cline = new Cline(this, apiConfiguration, customInstructions, alwaysAllowReadOnly, alwaysAllowWrite, alwaysAllowExecute, alwaysAllowBrowser, task, images)
```

- Others:
    - getStateToPostToWebview
    - vscode.WebviewViewProvider
    - ExtensionState
    - ExtensionStateContextType
    - ExtensionStateContextProvider
    - WebviewMessage
    - SettingsView 
        - include vscode "html"-like option rendering
        - should be dynamic based on configuration type. 
        - special configuration-specific attributes can be annotated in config_desc. these might include:
            - CSS and other configuration specific formatting considerations
            - position in the configuration interface (tabs, etc)
            - possibly other config attrs
    - and probably others...
    
        
Relatedly, it would be great to have per project configuration, in `{project-root}/.cline/config.json` or something like that. Then users can keep their Cline configurations in git with the project, especially since after projects require project-specific AI system instructions.

**Comment by lloydchang** [comment]
Related https://github.com/cline/cline/discussions/368 https://github.com/cline/cline/discussions/635 https://github.com/cline/cline/pull/674 https://github.com/cline/cline/pull/784 https://github.com/cline/cline/pull/802 https://github.com/cline/cline/discussions/835 https://github.com/RooVetGit/Roo-Cline/pull/1 https://github.com/RooVetGit/Roo-Cline/pull/14 https://github.com/lloydchang/cline-cline-fka-saoudrizwan-claude-dev/pull/20
